modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyTree.jsx

```diff
@@ -117,6 +127,28 @@ function VocabularyTree(props) {
     }
   }
 
+  let onAddOption = (name, path) => {
+    let newTerms = selectedTerms.slice();
+    newTerms.push([name, path]);
+    setSelectedTerms(newTerms);
+    // remove from removed
+    let newRTerms = removedTerms.filter(item => item[1] != path);
+    setRemovedTerms(newRTerms);
+  }
+
+  let onRemoveOption = (name, path) => {
+    let terms = selectedTerms.filter(item => item[1] == path);
+    if (terms.length > 0) {
+      let newTerms = selectedTerms.filter(item => item[1] != path);
+      setSelectedTerms(newTerms);
+    } else {
+      // they were selected before, add to removed
+      let newRTerms = removedTerms.slice();
+      newRTerms.push([name, path]);
+      setRemovedTerms(newRTerms);
+    }
```

Comment:
I'm not sure I follow the logic here. The comment says "they were selected before, add to removed". When is "before"? should the removed term be checked against `initialSelection` instead? The following seems to make more sense to me, but maybe I'm missing something.

```suggestion
    let newTerms = selectedTerms.filter(item => item[1] != path);
    setSelectedTerms(newTerms);
    if (initialSelection.some(item => item[1] == path)) {
      // they were selected before, add to removed
      let newRTerms = removedTerms.slice();
      newRTerms.push([name, path]);
      setRemovedTerms(newRTerms);
    }
```

Reply:
Addressed in 96823e0